Ir 507 Refactor InteractableComponent and expose API

SamMazerIR opened the pull request:
## Summary
Refactor Interactables to include their own callbacks and XRUI label information. standardized interatables nomenclature (from interactive) and fixed issues with mediacontrols and mount point not showing XRUI.

## Subtasks Checklist
[https://tsu.atlassian.net/browse/IR-569](IR-569)
[https://tsu.atlassian.net/browse/IR-521](IR-521)

## Breaking Changes


## References
closes #_insert number here_

## QA Steps
Create a test scene with an interactable, video and a mount point. interactable can reference any entities in the scene with callbacks using the editor UI

SamMazerIR committed:
giving media controls a proper location, fixing visibility issue with interactables XRUI and distance check logic for enabling them - now uses avatar rigidbody position vs a chest bone in the normalizedrig that was not being updated

SamMazerIR committed:
WIP refactoring interactables to work more general purpose. Interactions can now be set from the editor with a target entity and a callback to trigger on interact.
Still need to add an InteractUI automatically and use the InteractableComponent label for the text. (this will likely break MountPoint as it is written)

SamMazerIR committed:
Editor label input for XRUI of interactable now working, interactive references now refactored to be interactable for consistent naming. also updated some changelog references to match name change

SamMazerIR committed:
adding localization information

DanielBelmes commented:
@cla-bot check

SamMazerIR committed:
reworking a block of code so that an optional chain doesn't get flagged by code analysis as unsafe (even with an early return making it safe)

SamMazerIR committed:
Merge branch 'dev' into IR-507-Interactables

HexaField commented on packages/engine/src/interaction/components/properties/InteractableComponentNodeEditor.tsx:
```diff
@@ -0,0 +1,180 @@
+      </PropertiesPanelButton>
+
+      <div key={`callback-list-${props.entity}`}>
+        {interactableComponent.callbacks.map((callback, index) => {
```
Need to key with the index the return value

SamMazerIR replied:
addressed

SamMazerIR committed:
disabling interactableUI for editor mode

SamMazerIR committed:
making suggested changes for keys

SamMazerIR committed:
fixing an issue with onClick invocation

SamMazerIR committed:
fixing LinkComponent (bad input capture check) and refactoring MountPointSystem, also adding a captured by view check for AvatarCameraInputSystem and using as early return, changing to run after InputSystem instead of with. giving GrabbableComponent a node editor and i18n entries

TODO: possibly integrate InteractableSystem into LinkComponent, definitely integrate into GrabbableComponent

SamMazerIR committed:
minor cleanup

SamMazerIR committed:
WIP on grabbablesystem, if you force your way past the handedness and network issues you get expected behavior from the interactable now. it's a minor savings since it still tracks its own input for dropping

SamMazerIR committed:
adding a RigidBodyComponent on mount in GrabbableComponent as it is required by grabbable system

SamMazerIR committed:
Merge branch 'dev' into IR-507-Interactables

SamMazerIR committed:
removing comments in linkcomponent

HexaField commented on packages/engine/src/interaction/components/GrabbableComponent.ts:
```diff
@@ -33,7 +37,16 @@ import { Entity } from '@etherealengine/ecs/src/Entity'
+  reactor: function () {
+    const entity = useEntityContext()
+    useEffect(() => {
+      setComponent(entity, RigidBodyComponent, { type: BodyTypes.Kinematic })
```
we only want to make rigidbodies kinematic if when they are grabbed, and not do it if there is not a rigidbody component

SamMazerIR replied:
removing this logic completely, did not realize that grabbables are meant to function even in the absence of a RigidBodyComponent

HexaField commented on packages/engine/src/interaction/systems/GrabbableSystem.tsx:
```diff
@@ -377,7 +373,18 @@ const execute = () => {
    */
   if (isClient)
     for (const entity of grabbableQuery.enter()) {
```
lets move this enter query into a reactor mount

SamMazerIR replied:
moved this into the component instead on mount, did the same for the unmount logic

HexaField commented on packages/engine/src/interaction/systems/GrabbableSystem.tsx:
```diff
@@ -390,19 +397,26 @@ const execute = () => {
 const executeInput = () => {
-  const inputPointerEntity = InputPointerComponent.getPointerForCanvas(Engine.instance.viewerEntity)
-  if (!inputPointerEntity) return
+  // const inputPointerEntity = InputPointerComponent.getPointerForCanvas(Engine.instance.viewerEntity)
```
i we arent planning on using spatial input here, we should remove the commented code

SamMazerIR replied:
done, got rid of this as it didn't make sense with interactable distance checking anymore, and also the use case of XR hands being basically on top of a grabbable but perhaps rotated wrong (not pointing at it) felt like a miss if it fails to grab

HexaField commented on packages/engine/CHANGELOG.md:
```diff
@@ -216,7 +216,7 @@ See [Conventional Commits](https://conventionalcommits.org) for commit guideline
 ### Reverts
 
-* Revert "Fix some interactive system errors while scene is loading" ([126fef5](https://github.com/XRFoundation/XREngine/commit/126fef5e884e4c4cf7087ad0c49d7ba0b90ddca0))
+* Revert "Fix some interactable system errors while scene is loading" ([126fef5](https://github.com/XRFoundation/XREngine/commit/126fef5e884e4c4cf7087ad0c49d7ba0b90ddca0))
```
please revert this auto replace in the changelog history

SamMazerIR replied:
reverted

HexaField commented on packages/engine/src/avatar/systems/AvatarCameraInputSystem.ts:
```diff
@@ -192,6 +196,6 @@ const execute = () => {
 export const AvatarCameraInputSystem = defineSystem({
   uuid: 'ee.engine.AvatarCameraInputSystem',
-  insert: { with: InputSystemGroup },
+  insert: { after: InputSystemGroup },
```
why has this been changed to after?

SamMazerIR replied:
on a call we were trying to get the AvatarCameraInputSystem capture of input to happen last, it did not work, will revert

HexaField commented on packages/engine/src/interaction/components/InteractableComponent.ts:
```diff
@@ -23,6 +23,59 @@ All portions of the code written by the Ethereal Engine team are Copyright © 20
+    if (json.label) component.label.set(json.label)
+
+    // backwards compatibility
+    const callbackID = (json as any).callbackID ?? null
```
we don't need backwards compat outside the context of this PR, so lets remove this

SamMazerIR replied:
removed

HexaField commented on packages/engine/src/interaction/components/properties/InteractableComponentNodeEditor.tsx:
```diff
@@ -0,0 +1,180 @@
+
+import PanToolIcon from '@mui/icons-material/PanTool'
+
+// import { Button } from '../inputs/Button'
```
remove commented imports

SamMazerIR replied:
whups, done

HexaField commented on packages/engine/src/scene/components/MediaComponent.ts:
```diff
@@ -134,6 +134,7 @@ export const MediaComponent = defineComponent({
       controls: false,
       synchronize: true,
       autoplay: true,
+      uiOffset: new Vector3(0, 0, 0),
```
why are we redeclaring this here, if the media component can be composed with the interactable component for the UI?

SamMazerIR replied:
I added this as an optional offset for the specialized placement of the mediacontrols xrui, which do not use the default interactable flow as of right now. happy to discuss further or make changes (this was a light touch to first fix mediacomponent, but it uses its own custom update for the xrui for logic to show or hide...as well as custom positioning) might try to converge this further

AidanCaruso committed:
Merge branch 'dev' into IR-507-Interactables

AidanCaruso commented on packages/engine/src/interaction/systems/InteractableSystem.ts:
```diff
@@ -29,31 +29,39 @@ import { Euler, Quaternion, Vector3 } from 'three'
 
-export const InteractState = defineState({
-  name: 'InteractState',
+export const InteractableState = defineState({
```
Can we make this state return just the entity by distance array and put the max distance in the interactable component to make things more easily configureable?

SamMazerIR replied:
I like configurability but this may also cause confusion as to which interactable (IA) will fire (close to a normal range IA but within the threshold of an extended range IA that is further away...we will always activate the closer anyway)

SamMazerIR replied:
noting that this has been implemented with per-interactable distances that can be specified in editor. Also implemented with frustum culling

AidanCaruso commented on packages/engine/src/interaction/systems/GrabbableSystem.tsx:
```diff
@@ -372,11 +373,22 @@ const execute = () => {
+        ]
+      })
+
+      //addInteractableUI(entity, createInteractUI(entity, grabbableInteractMessage), onGrabbableInteractUpdate)
```
onGrabbableInteractUpdate callback is never used if this stays commented out. Is it beneficial to pass these update functions as callbacks? It seems to result in some duplicate code between the grabbable update function and the interactable update function.

SamMazerIR replied:
taken care of

AidanCaruso commented on packages/engine/src/interaction/systems/MountPointSystem.ts:
```diff
@@ -199,10 +203,10 @@ const vec3_1 = new Vector3()
   const mountedEntities = useHookstate(getMutableState(MountPointState))
   useEffect(() => {
-    //temporary logic for setting visibility of hints until interactive system is refactored
+    //temporary logic for setting visibility of hints until interactable system is refactored
```
This comment probably doesn't apply anymore, unless we want to add logic for hiding the hints on interact

SamMazerIR replied:
noting that i want to revisit how we handle hiding on Monday, as this is still currently required to hide as-written and if feels like a hack workaround

SamMazerIR replied:
no longer relevant with new system design

SamMazerIR committed:
moving grabbable functionality into shared grabbableFunctions, cleaning up based on requested changes, show/hide media controls on hover now - click video to play/pause...need to move play button to a better spot - currently the xrui is non interactable

SamMazerIR committed:
Merge branch 'IR-507-Interactables' of github.com:EtherealEngine/etherealengine into IR-507-Interactables

SamMazerIR committed:
adding license comment

SamMazerIR committed:
grab tests import fix

SamMazerIR committed:
adding custom distance per interactable as an exposed field, changing activation to fire for ALL interactables in range, separating MediaComponent/System from Interactables. Removing map of custom updates from MediaControlSystem - TODO remove map from InteractableSystem as well

SamMazerIR committed:
adding highlights to interactables as an option and cursor change if hover activated
Interactables now have optional clickability support (exposed in editor)
override state for UI activation - removes need for custom updates on XRUI for components using Interactables
folded LinkComponent, MountPointSystem, and GrabbableComponent/System  in to use InteractableComponent/System
removed nonInteractUI in favor of consolidating with createInteractUI which is now createUI (and now currently unused)
fix for mediaControlSystem not updating state reactively
fixed some editor i18n

SamMazerIR committed:
Merge branch 'dev' into IR-507-Interactables

# Conflicts:
#	packages/editor/src/functions/addMediaNode.ts - took dev

SamMazerIR committed:
linter issues, removing type annotation from inferred boolean type params with default values

the-reality-engineer reviewed (approved):
We talked about this over voice, but we can do this in the next PR; we should also introduce the concept of a focussed interactable Entity, such that it is the (only) entity that will respond to the interact button being pressed.

SamMazerIR committed:
media controls icon now updates properly, is positioned in the center of the media by default again (though I left the commented code for anchoring it to lower left outside the media box. I still feel this would be useful when combined with a choice of preset locations)

SamMazerIR committed:
adding frustum culling to interactables eligibility to activate and its XRUI being active or not. renamed file "gatherAvailableInteractables" to "interactableFunctions" to better reflect its new contents/usages

HexaField commented on packages/engine/src/interaction/components/properties/GrabbableComponentNodeEditor.tsx:
```diff
@@ -0,0 +1,47 @@
+Ethereal Engine. All Rights Reserved.
+*/
+
+import { useComponent } from '@etherealengine/ecs'
```
please move this to the editor package, this currently breaks the pacakge hierarchy

SamMazerIR replied:
moving

HexaField commented on packages/engine/src/interaction/components/properties/InteractableComponentNodeEditor.tsx:
```diff
@@ -0,0 +1,190 @@
+Ethereal Engine. All Rights Reserved.
+*/
+
+import { UUIDComponent } from '@etherealengine/ecs'
```
please move this to the editor package, this currently breaks the pacakge hierarchy

SamMazerIR replied:
moving

HexaField commented on CHANGELOG.md:
```diff
@@ -216,7 +216,7 @@ See [Conventional Commits](https://conventionalcommits.org) for commit guideline
 ### Reverts
 
-* Revert "Fix some interactive system errors while scene is loading" ([126fef5](https://github.com/xrfoundation/XREngine/commit/126fef5e884e4c4cf7087ad0c49d7ba0b90ddca0))
+* Revert "Fix some interactable system errors while scene is loading" ([126fef5](https://github.com/xrfoundation/XREngine/commit/126fef5e884e4c4cf7087ad0c49d7ba0b90ddca0))
```
this still isn't reverted

SamMazerIR replied:
fixing, don't know how I missed this

SamMazerIR committed:
reverting change log change

SamMazerIR committed:
moving grabbable and interactable node editors to editor/src/components/properties, adding interactable, grabbable, and screenshare to component/node editor list. adding missing i18n entries for rigidbody and triggers

the-reality-engineer commented on packages/engine/src/interaction/systems/InteractableSystem.ts:
```diff
@@ -0,0 +1,325 @@
+  }
+}
+
+const interactWithClosestInteractable = () => {
```
Just a note for the next PR; This should accept an input source as a parameter. The heuristic should be something like this: 

- Find the interactable that is closest to input source's target ray in the viewer's viewing frustum
  - This could be the closest interactable that intersects with that ray, or;
  - The closest interactable to that ray
- Run the interact callback on that interactable

SamMazerIR replied:
will create an issue, thanks

the-reality-engineer commented:
All requested changes have been addressed, so I'm merging this in!
